Merge 2.5.x into 2.6.x

alcaeus opened the pull request:
Should've done this before creating new pipeline stages, that would've helped avoid a nasty conflict resolution. Oh well 🤷‍♂️

franmomu committed:
Update branch metadata (#2508)

franmomu committed:
Rename abstract test classes to *TestCase (#2507)

franmomu committed:
Remove occurrences of "$this" in data providers

IonBazan committed:
Merge pull request #2510 from franmomu/phpunit/remove_this_data_providers

Remove occurrences of "$this" in data providers

franmomu committed:
Avoid triggering deprecation (#2511)

When calling UnitOfWork::clear() without argument

IonBazan committed:
Use static data providers (#2512)

* Use static data providers

* Update phpunit.xml.dist

Co-authored-by: Claudio Zizza <859964+SenseException@users.noreply.github.com>

---------

Co-authored-by: Claudio Zizza <859964+SenseException@users.noreply.github.com>

alcaeus committed:
Merge branch '2.5.x' into merge-2.5.x

* 2.5.x:
  Use static data providers (#2512)
  Avoid triggering deprecation (#2511)
  Remove occurrences of "$this" in data providers
  Rename abstract test classes to *TestCase (#2507)
  Update branch metadata (#2508)

alcaeus committed:
Update phpstan errors

alcaeus commented:
Not sure I understand the psalm failures, @franmomu could you please take a look?

franmomu commented:
The thing about `Builder::getPipeline()` is that it returns `PipelineExpression`

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php#L274

which is `list<StageExpression>`:

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php#L33

and `StageExpression`:

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Stage.php#L18

so in `Builder::getPipeline()`

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php#L289-L292

`$stage->getExpression()` should return `StageExpression` or `non-empty-array<non-empty-string, mixed>`, if that's the case, in

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/MatchStage.php#L19-L22

should return a `non-empty-array`, but `Expr::getExpression()` returns `array<string, mixed>`

https://github.com/doctrine/mongodb-odm/blob/3091159e4758a37dbe92a034cf8a26df2fb84c2d/lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php#L515-L519

franmomu committed:
Fix psalm build (#2525)

* Fix return type

* Fix psalm build

These new errors are because a bug in psalm with get_class
see https://github.com/vimeo/psalm/issues/9607, they will be
gone once we use ::class

franmomu committed:
Merge remote-tracking branch 'upstream/2.5.x' into merge-2.5.x

franmomu committed:
update SA baseline errors

alcaeus commented:
Closing in favour of #2527 where I've also fixed the issue above. Thanks for your help @franmomu!
